util/network-devp2p/src/enr.rs

```diff
@@ -0,0 +1,434 @@
+pub struct Record {
+    seq: u64,
+    signature: Option<Vec<u8>>,
+    raw: Option<Vec<u8>>, // RLP encoded record.
+    pairs: HashMap<String, Value>, // Sorted list of key/value pairs.
+}
+
+impl Record {
+    pub fn get(&self, entry: &dyn Entry) -> Option<&Value> {
+        self.pairs.get(&entry.enr_key())
+    }
+
+    /// Add or update a value for a key.
+    pub fn set(&mut self, value: &Value) {
+        self.invalidate();
+        let val = self.pairs.entry(value.enr_key()).or_insert(value.clone());
+        *val = value.clone();
+    }
+
+    fn invalidate(&mut self) {
+        if self.signature.is_some() {
+            self.seq += 1;
+        }
+
+        self.signature = None;
+        self.raw = None;
+    }
+
+    pub fn signature(&self) -> Option<Vec<u8>> {
+        self.signature.clone()
+    }
+
+    pub fn scheme(&self) -> Option<String> {
+        let id = Id("".to_owned());
+        let value = self.get(&id);
+        match value {
+            Some(Value::Id(id)) => Some(id.0.clone()),
+            _ => {
+                error!(target: "network", "Can't get scheme for ENR");
+                None
+            },
+        }
+    }
+
+    pub fn set_signature(&mut self, scheme: &dyn Scheme, sig: &Vec<u8>) {
+        if !scheme.verify(self, &sig) {
+            error!(target: "network", "Can't verify scheme");
+            return
+        }
+
+        self.signature = Some(sig.clone());
+        let raw = encode(self);
+        self.raw = Some(raw);
+    }
+
+    pub fn reset_signature(&mut self) {
+        self.signature = None;
+        self.raw = None;
+    }
+
+    pub fn set_seq(&mut self, seq: u64) {
+        self.seq = seq;
+        self.signature = None;
+        self.raw = None;
+    }
+
+    pub fn seq(&self) -> u64 {
+        self.seq
+    }
+}
+
+impl Encodable for Record {
+    fn rlp_append(&self, s: &mut RlpStream) {
+        s.begin_list(self.pairs.len() * 2 + 3)
+            .append(&self.signature)
+            .append(&self.seq)
+            .append(&self.raw);
+
+        for (key, value) in &self.pairs {
+            s.append(key).append(value);
+        }
+
+        if s.len() > ENR_MAX_SIZE {
+            error!(target: "network", "Encoded bytes exceed limit(300 bytes), length is {:?}", s.len())
+        }
+    }
+}
+
+impl Encodable for Value {
+    fn rlp_append(&self, s: &mut RlpStream) {
+        s.begin_list(1);
+
+        match self {
+            Value::Tcp(tcp) => s.append(&tcp.0),
+            Value::Tcp6(tcp6) => s.append(&tcp6.0),
+            Value::Udp(udp) => s.append(&udp.0),
+            Value::Udp6(udp6) => s.append(&udp6.0),
+            Value::Id(id) => s.append(&id.0),
+            Value::IpAddr(ip) => match ip {
+                IpAddr::V4(ip4) => s.append_raw(&ip4.octets()[..], 4),
+                IpAddr::V6(ip6) => s.append_raw(&ip6.octets()[..], 16),
+            }
+        };
+    }
+}
+
+/// Tcp port of a node
+#[derive(Clone, PartialEq, Debug)]
+pub struct Tcp(u16);
```

Comment:
All those probably should have `pub` modifiers or `Deref`/`AsRef` implementations. Otherwise the inner value will not be accessible.

Reply:
done